File: day23.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct cup_s {
  int  next;
  char moving;
} cup;

void play(char *input, int n, int turns)
{
  cup *cups = malloc((n + 1) * sizeof(cup));
  int  i, prev = 0, current = input[0] - '0';

  /* Load the seed */

  for(i = 0; i < strlen(input); i++)
  {
    int v = input[i] - '0';

    cups[v].moving = 0;
    
    if(prev)
      cups[prev].next = v;

    prev = v;
  }
  /* Fill out for part 2 */

  if(n > strlen(input))
  {
    for(i = strlen(input) + 1; i <= n; i++)
    {
      cups[prev].next = i;
      prev = i;
    }
  }
  cups[prev].next = current;

  for(i = 0; i < turns; i++)
  {
    int j, c = current, t = (c + n - 2) % n + 1, save;

    /* Mark the moving cups */

    for(j = 0; j < 3; j++)
    {
      c = cups[c].next;
      cups[c].moving = 1;
    }
    /* Save the next cup after the olds that are moving */ 

    save = cups[c].next;

    /* Find the target cup for the move */

    while(cups[t].moving)
      t = (t + n - 2) % n + 1;

    /* Splice */

    cups[c].next = cups[t].next;
    cups[t].next = cups[current].next;
    cups[current].next = save;
    c = t;

    /* Unmark moving cups */

    for(j = 0; j < 3; j++)
    {
      c = cups[c].next;
      cups[c].moving = 0;
    }
    current = save;
  }
  /* Read out the result */

  current = 1;

  if(n < 10)
  {
    printf("Part 1: ");

    for(i = 0; i < n; i++)
    {
      printf("%d", current);
      current = cups[current].next;
    }
    printf("\n");
  }
  else
  {
    unsigned long a = cups[1].next, b = cups[a].next, v = a * b;

    printf("Part 2: %d * %d = %lu\n", cups[1].next, cups[cups[1].next].next, v);
  }
}

int main(int argc, char *argv[])
{
  /* Arg checks */

  if(argc < 2)
  {
    printf("%s <cups>\n", argv[0]);
    exit(1);
  }
  play(argv[1], 9, 100);
  play(argv[1], 1000000, 10000000);
}

File: day23.rb
#!/usr/bin/env ruby

def play(input, n, turns)
  prev = 0
  cups = [0] * (n + 1)
  current = input[0].to_i

  for i in 1..(input.length) do
    cups[prev] = {'next': input[i - 1].to_i, 'moving': false} if prev > 0
    prev = input[i - 1].to_i
  end

  if n > input.length
    for i in (input.length + 1)..n do
      cups[prev] = {'next': i, 'moving': false}
      prev = i
    end
  end

  cups[prev] = {'next': current, 'moving': false}

  for i in 1..turns
    c = current
    t = (c - 2) % n + 1

    for j in 1..3 do
      c = cups[c][:next]
      cups[c][:moving] = true
    end

    save = cups[c][:next]
    m = cups[t][:moving]

    while m do
      t = (t - 2) % n + 1
      m = cups[t][:moving]
    end

    cups[c][:next] = cups[t][:next]
    cups[t][:next] = cups[current][:next]
    cups[current][:next] = save
    c = t

    for j in 1..3 do
      c = cups[c][:next]
      cups[c][:moving] = false
    end
    
    current = save
  end

  current = 1

  if n < 10
    output = ''
    for i in 1..n do
      output += current.to_s
      current = cups[current][:next]
    end
    puts "Part 1: #{output}"
  else
    puts "Part 2: #{cups[1][:next] * cups[cups[1][:next]][:next]}"
  end
end

play(ARGV[0], 9, 100)
play(ARGV[0], 1000000, 10000000)
